core/Pos/Crypto/SafeSigning.hs

```diff
@@ -72,33 +75,60 @@ emptyPassphrase = PassPhrase mempty
 instance Default PassPhrase where
     def = emptyPassphrase
 
-{-instance Monoid PassPhrase where
-    mempty = PassPhrase mempty
-    mappend (PassPhrase p1) (PassPhrase p2) = PassPhrase (p1 `mappend` p2)-}
-
-mkEncSecret :: Bi PassPhrase => PassPhrase -> CC.XPrv -> EncryptedSecretKey
-mkEncSecret pp payload = EncryptedSecretKey payload (hash pp)
+-- | Parameters used to evaluate hash of passphrase.
+passScryptParam :: S.ScryptParams
+passScryptParam =
+    fromMaybe (error "Bad passphrase scrypt parameters") $
+    S.mkScryptParams def
+        { S.spHashLen = 32  -- maximal passphrase length
+        }
+
+-- | Wrap raw secret key, attaching hash to it.
+-- Hash is evaluated using given salt.
+mkEncSecretWithSalt
+    :: Bi PassPhrase
+    => S.Salt -> PassPhrase -> CC.XPrv -> EncryptedSecretKey
+mkEncSecretWithSalt salt pp payload =
+    EncryptedSecretKey payload $ S.encryptPassWithSalt passScryptParam salt pp
+
+-- | Wrap raw secret key, attachind hash to it.
+-- Hash is evaluated using generated salt.
+mkEncSecret
+    :: (Bi PassPhrase, MonadRandom m)
+    => PassPhrase -> CC.XPrv -> m EncryptedSecretKey
+mkEncSecret pp payload =
+    EncryptedSecretKey payload <$> S.encryptPass passScryptParam pp
 
 -- | Generate a public key using an encrypted secret key and passphrase
 encToPublic :: EncryptedSecretKey -> PublicKey
 encToPublic (EncryptedSecretKey sk _) = PublicKey (CC.toXPub sk)
 
 -- | Re-wrap unencrypted secret key as an encrypted one
-noPassEncrypt :: Bi PassPhrase => SecretKey -> EncryptedSecretKey
-noPassEncrypt (SecretKey k) = mkEncSecret emptyPassphrase k
-
-checkPassMatches :: (Bi PassPhrase, Alternative f) => PassPhrase -> EncryptedSecretKey -> f ()
-checkPassMatches pp (EncryptedSecretKey _ pph) = guard (hash pp == pph)
-
-changeEncPassphrase
+noPassEncrypt
     :: Bi PassPhrase
+    => SecretKey -> EncryptedSecretKey
+noPassEncrypt (SecretKey k) =
+    mkEncSecretWithSalt (S.mkSalt ("" :: Text)) emptyPassphrase k
```

Comment:
This kind of salt seems to be insecure since its serialization is something really simple and predictable. I'm by no means a theoretical cryptographer, but stackoverflow suggests to use non-trivial salts. Why not put some `"kolbasa  selyodka shapka"` here instead? And tbh I would put the default salt into the separate top-level definition (even though it's only used here).

Reply:
@volhovM it actually doesn't matter at all which salt to use _here_, because here we provide empty password, which efficiently says "we don't want any security for this key"